lib/inferno/dsl/fhirpath_evaluation.rb

```diff
@@ -0,0 +1,135 @@
+        # @return [Array] An array representing the result of evaluating the given expression against
+        #   the given root resource. Each "result" in the returned array will be in the form
+        #   `{ "type": "[FHIR datatype name]", "element": [JSON representation of element] }`.
+        def evaluate_fhirpath(fhir_resource, fhirpath_expression, runnable)
+          begin
+            response = call_fhirpath_service(fhir_resource, fhirpath_expression)
+          rescue StandardError => e
+            # This could be a complete failure to connect (fhirpath service isn't running)
+            # or a timeout (fhirpath service took too long to respond).
+            runnable.add_message('error', e.message)
+            raise Inferno::Exceptions::ErrorInFhirpathException, "Unable to connect to FHIRPath service at #{url}."
+          end
+          return JSON.parse(response.body) if response.status.to_s.start_with? '2'
+
+          runnable.add_message('error', "FHIRPath service Response: HTTP #{response.status}\n#{response.body}")
+          raise Inferno::Exceptions::ErrorInFhirpathException,
+                'FHIRPath service call failed. Review Messages tab for more information.'
+        rescue JSON::ParserError
+          runnable.add_message('error', "Invalid FHIRPath service response format:\n#{response.body}")
+          raise Inferno::Exceptions::ErrorInFhirpathException,
+                'Error occurred in the FHIRPath service. Review Messages tab for more information.'
+        end
+
+        def call_fhirpath_service(fhir_resource, fhirpath_expression)
+          Faraday.new(
+            url,
+            request: { timeout: 600 }
+          ).post(
+            "evaluate?path=#{fhirpath_expression}",
+            fhir_resource.to_json,
+            content_type: 'application/json'
+          )
+        end
+      end
+
+      module ClassMethods
+        # @private
+        def fhirpath_evaluators
+          @fhirpath_evaluators ||= {}
+        end
+
+        # Find a particular fhirpath evaluator. Looks through a runnable's parents up to
+        # the suite to find an evaluator with a particular name
+        def find_fhirpath_evaluator(evaluator_name)
+          evaluator = fhirpath_evaluators[evaluator_name] || parent&.find_fhirpath_evaluator(evaluator_name)
+
+          raise Inferno::Exceptions::FhirpathNotFoundException, evaluator_name if evaluator.nil?
+
+          evaluator
+        end
+
+        # Define a fhirpath evaluator
+        # @example
+        #   fhirpath_evaluator url: 'http://example.com/fhirpath'
+        #
+        #   fhirpath_evaluator name: :custom, url: 'http://example.com/custom-fhirpath'
+        #
+        # @param name [Symbol] the name of the fhirpath evaluator, only needed if you are
+        #   using multiple evaluators
+        # @param url [String] the url of the fhirpath service
+        def fhirpath_evaluator(name: :default, url: nil)
```

Comment:
Is there a reason why test writers would need to have multiple fhirpath services? I was really hoping this would just work out of the box without having to define fhirpath services.

Reply:
updated